Ordering of videos in What's new Section

123vipulj opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

I added functionality of ordering videos  by minutes and hours .videos added automatically in  their respective place in real time .

![53789457-6b30c980-3f4a-11e9-9daf-4fa10f605970](https://user-images.githubusercontent.com/1891273/53808342-3d4f8300-3f52-11e9-8ff8-3f5311d14106.png)

TobiGr commented:
**Thanks for taking a look at it again!**
Unfortunately, this didn't work for me. What I did:

1. Install the app (no subscriptions)
1. Go to What's New: => no subscriptions. Works as expected
1. Subscribe to a channel which didn't upload a video within the last month
1. Go to What's New: => Endless loading spinner
1. Subscribe to a channel which uploaded a video within the last hours
1. Go to What's New: => no subscriptions.
1. Subscribe to a channel which uploaded a video within the last days
1. Go to What's New: => no subscriptions.

Btw. When you take a closer look at your screenshot, you'll notice that the first video is not in order.

theScrabi commented:
I am not sure about this anyway, we should really focus on proper subscription instead of just messing around with sorting by date, which will only partiality solve our problem.

123vipulj commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -55,6 +57,20 @@
+    List<InfoItem> daysList = new ArrayList<>();
+    List<InfoItem> weeksList = new ArrayList<>();
+
+    int count = 2;
```
at 329 .
` if (itemsLoaded.size()  == count)`

itemLoaded  add two youtube subcription channel url at a  time and increase by 2 element every time. so i set count = 2 . whenever itemloaded equal to count after 2 element ,list will repopulate and  count will increase by 2.

123vipulj replied:
videos order by minute and hour not by (minute,hours) number . I think it is unnecessary to load weeks and days old videos so, i removed it. @TobiGr Can you just look at again by loading your entire youtube subcription .

123vipulj replied:
i will fix today 4 and 6 option

123vipulj replied:
**ordering happening by minute ,hour not by (minute, hour) numbers.**

123vipulj commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -292,16 +308,41 @@ public void onSuccess(final ChannelInfo channelInfo) {
-                    requestFeed(1);
+                for (int i = 0; i < 15; i++) {
+
+                    //get atleast 8 relative items of source youtube channel
```
it is 15. forgot to replace 8

123vipulj commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -329,12 +370,12 @@ private void onDone() {
 
                 int loaded = requestLoadedAtomic.incrementAndGet();
-                if (loaded >= Math.min(FEED_LOAD_COUNT, subscriptionPoolSize)) {
+                if (loaded < 7) {
```
This code do nothing.

123vipulj replied:
unnecessary code i removed it

123vipulj commented:
@TobiGr I updated my NewPipe Repo found out that youtube date showing 'minute ago', 'minutes ago' ,'day ago','days ago','hour ago','hours ago' but its not happen in previous versions . That's why above things not working that you tested .it's now fixed.

TobiGr commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -281,27 +297,56 @@ public void onComplete() {
+
+                    //categorize into time
+                    if (!itemExists) {
+                        if (((StreamInfoItem) infoItem).getUploadDate().contains("minutes ago") || ((StreamInfoItem) infoItem).getUploadDate().contains("minutes ago")) {
```
I guess one of those was pretended to be `"minute ago"`.

TobiGr commented:
Thanks for fixing it. But please be aware of different translations of "minutes", "days" etc. when a user does not get an English version of the channel.

123vipulj commented:
It won't matter whatever language i change 
This code always return Time in english  in what's new and trending page. I think it universal throughout app except in searching videos.

`(StreamInfoItem) infoItem).getUploadDate()`

TobiGr commented:
> I think it universal throughout app except in searching videos.

No. this is exactly one of the problems we have. See #578 and #2085. For example for me in Germany, the `uploadDate` is `vor 36 Minuten` and therefore no videos are displayed in the What's New section.

theScrabi commented:
One solution to this problem is that you can temporarily request something in english even if default language difers. So for feed we could request alsway in english, and therefore focus on english translation.

theScrabi commented:
I want to take a look at this before merge, because i have the feeling parts of this commit belong into extractor

123vipulj commented:
@TobiGr issue #578 problem due to localization . if u set  content country to India it will show indonesia in first page of search videos after second page to English if content language set to english. It will be fine with some country content  and language set like as German and Deutche. It will fix in some day.

see video
 [https://drive.google.com/open?id=1rZgsAH8dzwC_DxEicOiNWEpQPHW3bp0B](url). 

One solution i found about what' new section to support multiple language to use shared preference related to localization then fetch translation(day,minute,hours,week) from string.xml  related with  language into **contains()** method . 
`(StreamInfoItem) infoItem).getUploadDate().contains("minutes ago").`

123vipulj commented:
issue #578  and #2085  fixed.

https://drive.google.com/open?id=1JOwP8iDKCvPs85T5ZD8nx0J026nONRb-

theScrabi commented:
@123vipulj pleae fix issue #578 in another pull request. Please only focus on one feature at a time.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -281,27 +297,56 @@ public void onComplete() {
-                    requestFeed(1);
+                for (int i = 0; i < 15; i++) {
+
+                    //get atleast 15 relative items of source youtube channel
```
Put the description here to where the constant is defined.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -281,27 +297,56 @@ public void onComplete() {
+
+                    //categorize into time
+                    if (!itemExists) {
+                        if (((StreamInfoItem) infoItem).getUploadDate().contains("minute ago") || ((StreamInfoItem) infoItem).getUploadDate().contains("minutes ago")) {
```
This only works on youtube, but it will not work with other services.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -328,21 +373,21 @@ private void onDone() {
-                    requestLoadedAtomic.set(0);
-                    isLoading.set(false);
-                }
+//                int loaded = requestLoadedAtomic.incrementAndGet();
```
Don't uncomment code, remove it if it's not in use any more.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -328,21 +373,21 @@ private void onDone() {
-                    isLoading.set(false);
-                }
+//                int loaded = requestLoadedAtomic.incrementAndGet();
+//                if (loaded < 7) {
```
Magic number. Why 7?

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -328,21 +373,21 @@ private void onDone() {
-                    if (infoListAdapter.getItemsList().size() == 0) {
-                        showEmptyState();
-                    }
+//                    if (infoListAdapter.getItemsList().size() == 0) {
```
Again remove code dont uncomend it.

theScrabi commented:
@TobiGr I'm still a fan of proper solutions that might take some more time but then work well, rather that hacky ones that work now but cause trouble later. I promise to take on this problem with the feed/sugestions this year, however I would want to make it in the right way. With the timeago parser there, and the temporary language change there has already been prework on this problem we just had to make it finally work. Give it some time please.

TobiGr commented:
> I don't know, this pull request only works on youtube, but not on other services. I get the feeling that we should make this work on the extractor side first before we can do something here.

I agree with you. The more I look into this, the more I have the feeling we need a proper solution _for all services_, too.

> With the timeago parser there, and the temporary language change there has already been prework on this problem we just had to make it finally work. Give it some time please.

This sounds like the right point to start from. We are all looking forward to this. 

Nevertheless, @123vipulj thank you for taking a look and spending the time and effort 👍
